### PR TITLE
WFLY-6603 Infinispan subsystem does not expose a proper capability contract

### PR DESCRIPTION
**Pull request by pferraro**
https://issues.jboss.org/browse/WFLY-6603

Capability support for clustering subsystems, phase 2 of 3:
~~Phase 1: JGroups subsystem~~
**Phase 2: Infinispan subsystem**
Phase 3: Clustering API

### CODE REVIEW

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/KeyAffinityServiceFactoryBuilder.java
Comment (bstansberry) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/KeyAffinityServiceFactoryBuilder.java]: Was this change from 10 to 100 deliberate?

I only ask because it came in in a refactoring commit so may have been accidental.

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLReader.java
Comment (bstansberry) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLReader.java]: What's this about? Why aren't they "batched"? (IOW is there some problem in the kernel?)

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/KeyAffinityServiceFactoryBuilder.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/KeyAffinityServiceFactoryBuilder.java]: It was Intentional, but should be in a separate commit.

Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/KeyAffinityServiceFactoryBuilder.java]: Done.

File: clustering/singleton/extension/src/main/java/org/wildfly/extension/clustering/singleton/SingletonPolicyResourceDefinition.java
Comment (bstansberry) [inline clustering/singleton/extension/src/main/java/org/wildfly/extension/clustering/singleton/SingletonPolicyResourceDefinition.java]: Are CACHE_CONTAINER and CACHE model reference attributes? They don't seem to be used as such but naive intuition makes it seem like they would be.

**Review by bstansberry — COMMENTED** [review]
Thanks for breaking this up. I don't claim to follow everything done here. :) But I definitely feel that on each of these commits I can follow the basic pattern and that if something didn't seem right there was a decent chance I'd notice.

So, minor points I'm posting here aside, @kabir this looks good to me. 

@pferraro there will be the JBEAP and 7.x PR hoop to jump through too; I'll leave that to you and Kabir. :)

File: clustering/singleton/extension/src/main/java/org/wildfly/extension/clustering/singleton/SingletonPolicyResourceDefinition.java
Comment (pferraro) [inline clustering/singleton/extension/src/main/java/org/wildfly/extension/clustering/singleton/SingletonPolicyResourceDefinition.java]: Yes - but this will be addressed in phase 3 (see the PR description), including:
- capability references for these attributes
- disallow expressions (because these are model references) + appropriate rejection transformation

File: clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLReader.java
Comment (pferraro) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLReader.java]: Kabir can provide some context - as I made this change based on his advise (we did the same for the jgroups changes from last week).  Essentially, this was done solely to make the subsystem test framework happy, since certain tests executes the operations created by the parser one at a time, rather than as a batch - and this will fail unless we order them like this.

Comment (bstansberry) [inline clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/InfinispanSubsystemXMLReader.java]: I thought about it later and my guess was it had something to do with subsystem tests. Do I win a prize?

This is definitely something we need to clean up as having this kind of thing proliferate through the codebase will be awful. But that's not this PR's problem.

File: clustering/singleton/extension/src/main/java/org/wildfly/extension/clustering/singleton/SingletonPolicyResourceDefinition.java
Comment (bstansberry) [inline clustering/singleton/extension/src/main/java/org/wildfly/extension/clustering/singleton/SingletonPolicyResourceDefinition.java]: OK, sounds good. I'm not sure transformation will be required as we won't allow the expression on the DC so the slave will just see a normal value.

**Review by bstansberry — APPROVED** [review]

**Comment by bstansberry** [comment]
retest this please

**Comment by bstansberry** [comment]
I removed "Ready for Merge" label because this is ok but is going to need a rebase once I merge other changes that are also ready to go. I think it will be easier/safer for you to resolve the conflict than the other side. Sorry! (It's just a 1 file conflict.)

**Comment by bstansberry** [comment]
@pferraro the other changes are in so this can be rebased now.

**Comment by pferraro** [comment]
@bstansberry Done.

**Comment by bstansberry** [comment]
retest this please

**Pull request merged**
